core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -69,20 +108,52 @@ class KafkaRequestHandler(id: Int,
           completeShutdown()
           return
 
+        case callback: RequestChannel.CallbackRequest =>
+          try {
+            val originalRequest = callback.originalRequest
+            
+            // If we've already executed a callback for this request, reset the times and subtract the callback time from the 
+            // new dequeue time. This will allow calculation of multiple callback times.
+            // Otherwise, set dequeue time to now.
+            if (originalRequest.callbackRequestDequeueTimeNanos.isDefined) {
+              val prevCallbacksTimeNanos = originalRequest.callbackRequestCompleteTimeNanos.getOrElse(0L) - originalRequest.callbackRequestDequeueTimeNanos.getOrElse(0L)
+              originalRequest.callbackRequestCompleteTimeNanos = None
+              originalRequest.callbackRequestDequeueTimeNanos = Some(time.nanoseconds() - prevCallbacksTimeNanos)
+            } else {
+              originalRequest.callbackRequestDequeueTimeNanos = Some(time.nanoseconds())
+            }
+            
+            currentRequest.set(originalRequest)
+            callback.fun()
+            if (originalRequest.callbackRequestCompleteTimeNanos.isEmpty)
```

Comment:
It is not true if we returned a response. We also update the value there.

Reply:
There are two cases according to artem --
one where we send a response via network client and one where we don't (and we have more callbacks)

in case a, we set in the network client and that is the end of handling of the request. This is also when the metric is updated. In case b, we don't set it there so we must do it here. I have the check to match the same sort of protocol we see in kafka apis where we check if the value is -1. Maybe it's fine to set it twice in case a as we won't update the metrics again, but I did it for consistency.

Reply:
I am still a bit confused on this.

case a - only 1 callback. In this case, callbackRequestCompleteTimeNanos is never set and is expected to be empty.

case b - multiple callbacks. In this case, after the first callback, callbackRequestDequeueTimeNanos is set. So, when the 2nd callback is processed, we set callbackRequestCompleteTimeNanos to empty before executing the callback. When we get here, callbackRequestCompleteTimeNanos is expected to be empty.

So, it seems that in both case a and b, we expect callbackRequestCompleteTimeNanos to be empty?

Reply:
Basically, we should search for `apiLocalCompleteTimeNanos` and update the callbackRequestCompleteTimeNanos in similar places.  My understanding is that we can send the reply during request / callback and it may complete (concurrently) before we end the request / callback or after we end request / callback.  In which case we want the end time to be the earliest of the two.
Usually, the way it's done is the start time is set at the start of the request / callback processing and the end time is set after, but if the end time is not set then the metric is reported, then we just report the current time.

Reply:
In case a, we set it in the network client
in case b, it is empty and we set it. Then, when executing the next callback, we take this time and subtract it from the start of the new callback. This gives us total time of all.

Reply:
Thanks for the explanation. I understand it now. callbackRequestCompleteTimeNanos can be non-empty if the response is sent before the callback completes. Currently, we update the metric when the sending of the response completes. So, a more accurate place to set callbackRequestCompleteTimeNanos in RequestChannel is when the response send completes. 

An alternative approach is to delay the updating of the request time metrics until the callback completes. This simplifies the setting of callbackRequestCompleteTimeNanos since it only needs to be done in KafkaRequestHandler and is more accurate since we don't need to cut off callbackRequestCompleteTimeNanos by response send complete time.

Both of these could be addressed in a followup jira.

Reply:
I'm not sure I follow the first part since right now, we need to account for all the callback local times. Unless we change to a cumulative callback total time where we periodically add to it, we will need to update either on the send OR when the callback completes (if we don't send yet)

I'm also not sure about waiting until the callback completes because that is after we send via the network client and that is no longer considered "local time"

Reply:
For simplicity, let's just assume there is only one callback for now. Currently, the code sets callbackRequestCompleteTimeNanos when the response is being sent. If the callback has been started, but not completed at that point, the measurement of the callback portion time is not accurate. If we delay the cut off of callbackRequestCompleteTimeNanos at the time when request time metric is updated (which currently is when the response send completes), we allow for more accurate measurement of the callback portion time.

For your second question, yes, it's a bit weird to report the callback time as 'local' after the response is being sent. However, 'local' really reflects the amount of time a request handler thread is tied up for processing a request. From that perspective, it could also make sense. If we cut off the callback portion of the time by response send time, we could leave a portion of the request thread time unaccounted for?

Reply:
I asked to just assume one callback earlier but I was told we should just implement this now 😅 

I'm not sure I follow 
> If the callback has been started, but not completed at that point, the measurement of the callback portion time is not accurate.

which scenario is inaccurate?

Reply:
In the more general case, the callback is not guaranteed to be completed when the response is sent, right? In that case, the code in RequestChannel cuts off callbackRequestCompleteTimeNanos at response send time. When the callback completes, KafkaRequestHandler doesn't set callbackRequestCompleteTimeNanos since it's not empty. This means the measurement of callback time is not accurate, right?

Reply:
Given that we have only one case right now I guess I'm not sure we have a general case -- but in the current case, we complete the callback by sending the response.

I guess I saw this protocol working as follows:

a) we have one callback and send the response in which local time is correct
b) we have multiple callbacks and the first few do not send the response. We updated after calling the callback. Then the final one sends the response (case a)

We always must return in the final callback because we only use this callback if we've already returned from handle. We must return via the callback in my understanding.

Reply:
Yes, in the current usage where the last callback sends the response as the last thing, it doesn't make much difference.

I was thinking that the callbacks in ActionQueue are executed after the response is generated. If want to replace ActionQueue with the callback queue, we need to handle the more general case.

Reply:
Yeah -- I guess the previous decisions were made without considering the action queue. 
I'm wondering if it is better to leave as is and change via a followup or simplify back to "base case"  (ie, case a which is currently what we do)

Reply:
https://issues.apache.org/jira/browse/KAFKA-14899